src/CarbonAware.LocationSources/src/LocationSource.cs

```diff
@@ -0,0 +1,137 @@
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _configurationMonitor = monitor ?? throw new ArgumentNullException(nameof(monitor));
+    }
+
+    public async Task<Location> ToGeopositionLocationAsync(Location location)
+    {
+        switch (location.LocationType)
+        {
+            case LocationType.Geoposition:
+            {
+                return await Task.FromResult(location);
+            }
+            case LocationType.CloudProvider: 
+            {
+                var geoPositionLocation = await GetGeoPositionLocationOrThrowAsync(location);
+                return geoPositionLocation;
+            }
+        }
+        throw new LocationConversionException($"Location '{ location.LocationType }' cannot be converted to Geoposition. ");
+    }
+
+    private async Task<Location> GetGeoPositionLocationOrThrowAsync(Location location)
+    {
+        await LoadRegionsFromFileIfNotPresentAsync();
+
+        var regionName = location.RegionName ?? string.Empty;
+        if (!_namedGeopositions!.ContainsKey(regionName))
+        {
+            throw new ArgumentException($"Unknown region: Region name '{regionName}' not found");
+        }
+
+        NamedGeoposition geopositionLocation = _namedGeopositions![regionName];    
+        if (!geopositionLocation.IsValidGeopositionLocation())  
+        {
+            throw new LocationConversionException($"Lat/long cannot be retrieved for region '{regionName}'");
+        }
+        if (_logger.IsEnabled(LogLevel.Debug))
+        {
+            _logger.LogDebug($"Converted Azure Location named '{location.RegionName}' to Geoposition Location at latitude '{location.Latitude}'" 
+                                + "and logitude '{longitude}'.", location.RegionName, geopositionLocation.Latitude, geopositionLocation.Longitude);
+        }
+        Location geoPosistionLocation = new Location 
+                {
+                    LocationType = LocationType.Geoposition,
+                    Latitude = Convert.ToDecimal(geopositionLocation.Latitude),
+                    Longitude = Convert.ToDecimal(geopositionLocation.Longitude)
+                };
+
+        return geoPosistionLocation;
+    }
+
+    protected virtual async Task<Dictionary<String, NamedGeoposition>> LoadRegionsFromJsonAsync()
+    {
+        var result = new Dictionary<String, NamedGeoposition>();
+        if (_configuration.LocationDataSources is null || !_configuration.LocationDataSources.Any())
+        {
+            _logger.LogInformation($"Loading default location data source");
+            await PopulateRegionMapAsync(result, new LocationDataSource());
+            return result;
+        }
+        foreach (var data in _configuration.LocationDataSources!)
+        {
+            await PopulateRegionMapAsync(result, data);
+        }
+        return result;
+    }
+
+    private async Task PopulateRegionMapAsync(Dictionary<String, NamedGeoposition> map, LocationDataSource data)
```

Comment:
Rather than plumbing the map all the way through, consider just updating the member variable directly from here.  It will clean up a bunch of code along the way.

`_namedGeopositions.Add(BuildKeyFromRegion(data, region), region);`